README.md

```diff
@@ -2,142 +2,295 @@
+- **Document size limit**: Documents larger than 1MB use `set` operations
+- **Change ratio threshold**: If >40% of text changes, uses `set` (indicates replacement vs. editing)
+- **Small document optimization**: Documents <10KB always use diff-match-patch
+- **System key protection**: Properties starting with `_` always use `set` operations
 
-To tune these rules, you can pass options to the differ:
+**Performance rationale:**
 
-```js
-import {diffPatch} from '@sanity/diff-patch'
+These thresholds are based on performance testing of the underlying `@sanity/diff-match-patch` library on an M2 MacBook Pro:
+
+- **Keystroke editing**: 0ms for typical edits, sub-millisecond even on large documents
+- **Small insertions/pastes**: 0-10ms for content <50KB
+- **Large insertions/deletions**: 0-50ms for content >50KB
+- **Text replacements**: Can be 70ms-2s+ due to algorithm complexity
+
+The 40% change ratio threshold catches problematic replacement scenarios while allowing the algorithm to excel at insertions, deletions, and small edits.
 
-diffPatch(itemA, itemB, {
-  diffMatchPatch: {
-    // Default is true, set to false to _always_ use `set`-patches
-    enabled: true,
-
-    // Only use diff-match-patch if target string is longer than this threshold
-    lengthThresholdAbsolute: 30,
-
-    // Only use generated diff-match-patch if the patch length is less than or equal to
-    // (targetString * relative). Example: A 100 character target with a relative factor
-    // of 1.2 will allow a 120 character diff-match-patch. If larger than this number,
-    // it will fall back to a regular `set` patch.
-    lengthThresholdRelative: 1.2,
-  },
-})
+**Migration from v5:**
+
+Version 5 allowed configuring diff-match-patch behavior with `lengthThresholdAbsolute` and `lengthThresholdRelative` options. Version 6 removes these options in favor of tested defaults that provide consistent performance across real-world editing patterns.
+
+### Array Handling
+
+**Keyed arrays**: Arrays containing objects with `_key` properties are diffed by key rather than index, producing more stable patches for collaborative editing.
+
+**Index-based arrays**: Arrays without keys are diffed by index position.
+
+**Undefined values**: When `undefined` values are encountered in arrays, they are converted to `null`. This follows the same behavior as `JSON.stringify()` and ensures consistent serialization. To remove undefined values before diffing:
+
+```js
+const cleanArray = array.filter((item) => typeof item !== 'undefined')
 ```
 
-In addition to these rules, there are certain cases where it will never use diff-match-patch:
+### System Keys
+
+The following top-level document keys are ignored when diffing as they are managed by Sanity:
 
-- If the patched key starts with an underscore (eg `_type`, `_key`, `_ref`)
-- If the diff-match-patch implementation cannot generate a legal patch due to unicode issues
```

Comment:
Did we remove this logic? I still feel we should always use set operations for `_type`, `_key`, `_ref` etc - even when not at the root. Patches becomes hard to read/reason about as diff-match-patch.

Reply:
yes, the logic is to use a `set` operation if the last path starts with `_`. this is also in the updated readme 😄